### ISSUE TITLE
How you can extract label probability scores from yolov8?

### ISSUE THREAD

**Issue by camiloromers** [ultralytics/ultralytics#4895]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello all,

I am trying to develop some active learning strategies but I need to get class label probabilities and after running predictions doing the following:

```python
results = model.predict(source="img.jpg", line_width=1, show_conf=False, save=True) 
```
I cannot access to any probability method in results.

Does anyone know how to do it? What I want is to get all probabilities for the same object such as:

class_1   prob 0.7
class_2  prob 0.2
class_3  prob 0.1

**Comment by RizwanMunawar** [comment]
@camiloromers Since you already possess the results array, you can utilize it for additional processing by referencing the provided example below.
```
results = model.predict(source="img.jpg", line_width=1, show_conf=False, save=True) 

boxes = results[0].boxes.xywh.cpu()                        #xywh bbox list
clss = results[0].boxes.cls.cpu().tolist()                 #classes Id list
names = results[0].names                                   #classes names list
confs = results[0].boxes.conf.float().cpu().tolist()       #probabilities of classes

for box, cls, conf in zip(boxes, clss, confs):
    x, y, w, h = box
    label = str(names[cls] + " "+ str(conf))
    x1, y1, x2, y2 = x-w/2, y-h/2, x+w/2, y+h/2            #xyxy bbox
    .......
......
```

**Comment by camiloromers** [comment]
> boxes = results[0].boxes.xywh.cpu()                        #xywh bbox list
> clss = results[0].boxes.cls.cpu().tolist()                 #classes Id list
> names = results[0].names                                   #classes names list
> confs = results[0].boxes.conf.float().cpu().tolist()       #probabilities of classes

Thanks @RizwanMunawar for your answer. That's partially what I want, but I would like to get all class probabilities for each bounding box. Let's say you are trying to detect 3 classes with yolo and but each predicted bonding box I'd like to get all class probabilities.

BBox 1 - class_1 `prob`, class_2 `prob`, class_3 `prob`
BBox2 - class_1 `prob`, class_2 `prob`, class_3 `prob`

and so on

**Comment by glenn-jocher** [comment]
@camiloromers it appears you're interested in not just the highest probability class for each bounding box, but the probabilities of each class. YOLOv8, by default, assigns each bounding box with the class associated with the highest probability score from the softmax output. 

However, to get all class probabilities for a bounding box, you would need to modify and access the layers in the model where class probabilities are decided. This layer is typically the output of the Softmax function, which gives the probabilities of each class.

Indeed, it is a deep dive into the source code and requires a fundamental understanding of the workings of the YOLOv8 model and PyTorch-based model layer interaction. If you're comfortable with model editing and understand the repercussions of making such a change, then this would be the way forward. For an operation like this, specific coding and deep learning model modification expertise is required. 

I hope that clarifies your question, and remember that while this modification is theoretically possible, it's not typically done nor is it necessarily straightforward. Please be aware of the risks and challenges this presents if you decide to proceed.

**Comment by camiloromers** [comment]
> @camiloromers it appears you're interested in not just the highest probability class for each bounding box, but the probabilities of each class. YOLOv8, by default, assigns each bounding box with the class associated with the highest probability score from the softmax output.
> 
> However, to get all class probabilities for a bounding box, you would need to modify and access the layers in the model where class probabilities are decided. This layer is typically the output of the Softmax function, which gives the probabilities of each class.
> 
> Indeed, it is a deep dive into the source code and requires a fundamental understanding of the workings of the YOLOv8 model and PyTorch-based model layer interaction. If you're comfortable with model editing and understand the repercussions of making such a change, then this would be the way forward. For an operation like this, specific coding and deep learning model modification expertise is required.
> 
> I hope that clarifies your question, and remember that while this modification is theoretically possible, it's not typically done nor is it necessarily straightforward. Please be aware of the risks and challenges this presents if you decide to proceed.

@glenn-jocher  that's exactly what I want. Do u have an example how I can get all probabilities from softmax or a layer behind?

**Comment by glenn-jocher** [comment]
Hello @camiloromers,

Extracting all class probabilities from the Softmax layer involves digging into the model architecture and data flow. It's not quite straightforward but let's break it down.

YOLOv8, much like other YOLO architectures, consists of several layers and modules, and the latter stages of the architecture usually involve a Softmax function. This function normalizes the output of the neural network to a probability distribution over predicted output classes. By default, each detected bounding box is assigned the class that has the highest probability from this Softmax function.

Since you intend to get the class probabilities for all classes and not just the highest, you'd have to modify the architecture slightly. Concretely, you need to access the output of the Softmax function where these class probabilities are being decided.

To implement this, you need a deep grasp of PyTorch and the workings of the YOLOv8 model, as it would involve rewriting parts of the model code and interacting with inner model layers, which isn't a standard operation. Furthermore, any modifications to the output might have unpredictable effects on the performance of the model.

I hope that helps, and do remember to tread carefully if you intend to follow through with this. Understanding what each layer does and how it affects the overall model is crucial in making effective modifications.

**Comment by camiloromers** [comment]
> Hello @camiloromers,
> 
> Extracting all class probabilities from the Softmax layer involves digging into the model architecture and data flow. It's not quite straightforward but let's break it down.
> 
> YOLOv8, much like other YOLO architectures, consists of several layers and modules, and the latter stages of the architecture usually involve a Softmax function. This function normalizes the output of the neural network to a probability distribution over predicted output classes. By default, each detected bounding box is assigned the class that has the highest probability from this Softmax function.
> 
> Since you intend to get the class probabilities for all classes and not just the highest, you'd have to modify the architecture slightly. Concretely, you need to access the output of the Softmax function where these class probabilities are being decided.
> 
> To implement this, you need a deep grasp of PyTorch and the workings of the YOLOv8 model, as it would involve rewriting parts of the model code and interacting with inner model layers, which isn't a standard operation. Furthermore, any modifications to the output might have unpredictable effects on the performance of the model.
> 
> I hope that helps, and do remember to tread carefully if you intend to follow through with this. Understanding what each layer does and how it affects the overall model is crucial in making effective modifications.

Thanks for your answer @glenn-jocher. Could you point me where exactly is the softmax function is please to see if I can work a little it and modify the code?

**Comment by glenn-jocher** [comment]
Hello @camiloromers,

The Softmax function is located deep within the structure of the YOLOv8 model. It is a part of the final stages of processing in the neural network, where it normalizes the output into a probability distribution over the predicted output classes. 

For a more specific location, you will need to look at the architecture definition of the YOLOv8 model. The output layer or computation that directly precedes the final prediction stage would likely be where the Softmax function, or an equivalent normalization process, is applied.

However, it is important to note that tweaking these sections of the model should be performed with great caution. It's a complex operation that implies potential risks as it could affect the performance and stability of the model.

I would recommend that you have a strong understanding of both PyTorch and the YOLOv8 model structure before proceeding. If you do decide to modify the code, ensure that you fully understand the implications and potential effects on the model's performance. 

I hope this response helps to give you an understanding of the area you might want to focus on. Best of luck with your exploration!
